### ISSUE TITLE
Allow SetupShop hook to work with AppliesToEntity hook

### ISSUE THREAD

**Issue by stormytuna** [bluemagic123/tModLoader#3152]
### Do you intend to personally contribute/program this feature?

No

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

Currently `SetupShop` will not take into account `AppliesToEntity` for the current NPC, so you can have a class like below and it will still add the Terra Blade to every npcs shop

```
    public class SomeGlobalNPC : GlobalNPC {
        public override bool AppliesToEntity(NPC entity, bool lateInstantiation) => entity.type == NPCID.Dryad;

        public override void SetupShop(int type, Chest shop, ref int nextSlot) {
            shop.item[nextSlot].SetDefaults(ItemID.TerraBlade);
            nextSlot++;
        }
    }
```

### What does this proposal attempt to solve or improve?

This would prevent issues where someone adds an item to a shop thinking it's only in that shop as they used `AppliesToEntity`. It would also prevent needing extra logic inside `SetupShop` if `AppliesToEntity` is being used for the rest of the code in that `GlobalNPC`

**Comment by Chicken-Bones** [comment]
Hi. Firstly I'd like to apologise for the confusion, perhaps we could add this to the documentation.

Unfortunately we can't do this in the current design, because `AppliesToEntity` takes an `NPC` parameter, but `SetupShop` doesn't have a full `NPC` instance, only a `int type`. Getting the actual `NPC` instance for all shops would be a significantly difficult change given the Terraria code.

**Comment by DavidF-Dev** [comment]
``NPC instance = Main.npc[Main.LocalPlayer.talkNPC]``
Would this be a possible solution for this issue? Depends if ``SetupShop`` is called in cases other than player interaction. I just helped a user in the Discord who wanted multiple instances of the same NPC type but with different shops; and this worked for them. They could access the specific instance being talked to.

If you're amending the documentation for ``GlobalNPC``, can I also request that you change the method description in ``ModNPC``; so that it's obvious it **isn't** being called on the instance of the npc being talked to.

For this reason, I think that [Example Traveling Merchant](https://github.com/tModLoader/tModLoader/blob/1.4/ExampleMod/Content/NPCs/ExampleTravelingMerchant.cs) does not work correctly. ``CreateNewShop`` is called each time a new instance is spawned, in ``SetDefaults``; however ``SetupShop`` will only use the shop generated for the dummy instance held in ``NPCLoader``. The generated shops of any other instance are not used; as can be seen in the below screenshots. Each screenshot is from a newly spawned instance of the NPC after killing the previous one in lava.

![2022-12-29 11_57_27-Window](https://user-images.githubusercontent.com/58208693/209890315-9e3fffdd-781b-42ef-b786-1cda77781241.png)
![2022-12-29 11_59_51-Window](https://user-images.githubusercontent.com/58208693/209890316-f4f7859b-2818-4b4a-8f81-9691a8364ffd.png)
![2022-12-29 12_00_25-Window](https://user-images.githubusercontent.com/58208693/209890318-059184c2-3ef3-4d61-b9df-b26a0e93b536.png)

Proposed fix:
```cs
public override void SetupShop(Chest shop, ref int nextSlot) {
	ExampleTravelingMerchant instance = (ExampleTravelingMerchant)Main.npc[Main.LocalPlayer.talkNPC].ModNPC;
	foreach (Item item in instance.shopItems) {
		// ...
	}
}
```

(Hopefully it's appropriate to leave my comment on this issue. If not, I can create a new issue.)

**Comment by Chicken-Bones** [comment]
Ultimately it's up to modders if they think it's more useful to have shops be bound to a specific instance of an NPC, or just an id. I'm reluctant to make a decision here because of:
- #3072

**Comment by DavidF-Dev** [comment]
That's fair. At the very least, I think it would be beneficial if the documentation/method description was updated to make it more obvious to the modder that it ISN'T being called on the instance being talked to; but rather a 'dummy' instance. As well as describing how they could get the correct instance if needed.

Another option may be to rename the method:
``SetupShop`` -> ``SetupShopOnType``/``StaticSetupShop``
Then introduce another method which is actually called on the correct instance:
``SetupShopOnInstance``/``SetupShop``
So that the modder can choose which approach they want based on which method they override (or both...).
This probably wouldn't mesh well with the "NPC Shop Modernisation" changes however.

Just to be clear, I wasn't suggesting that modders are forced to have shops bound to anything in particular; I was just stating that it's not obvious to most people unless they dive into the source code to see where exactly ``SetupShop`` is called from; and only then do they see that it isn't called on the *obvious* instance. The Traveling Merchant issue I raised is a bug in Example Mod, and the fix I suggested is for that specific example. I doubt it was intentionally written like that as that isn't how a Traveling Merchant should behave.
